### PR TITLE
Report exit status or signal that killed the test

### LINKED ISSUE #3642: When a test failures due to termination by a signal report that in the log
I've been helping bring the [Korn shell](https://github.com/att/ast/) project into the 21st century. It is not unusual for a unit test to fail, with no output, due to the ksh binary dying from a signal (e.g., SIGSEGV). Meson correctly reports the test failed but the details in *meson-logs/testlog.txt* is ambiguous. It would greatly help if details about the unit test exit status were reported in the log. For example, "Test terminated with exit status 3" or "Test terminated due to signal SIGSEGV". If the test passes there is no need to report "Test terminated with exit status 0".

### PR DESCRIPTION
**Pull request by krader1961**
When a test fails due to a signal (e.g., SIGSEGV) it can be somewhat
mysterious why the test failed. Also, even when a test fails due to a
non-zero exit status it would help if the exit status was reported. This
augments the result string to include the non-zero exit status or
signal number and name.

Resolves #3642

### CODE REVIEW

File: mesonbuild/mtest.py [hunk 1]
```diff
@@ -405,11 +405,32 @@ def print_stats(self, numlen, tests, name, result, i):
...
+            # process killed by a signal typically has a returncode < 128 while
+            # a termination due to an `exit()` call has a returncode > 128. So
+            # assume the usual UNIX semantics as munged by subprocess.
+            if result.returncode < 128:
+                status = ' (exit status %d)' % (result.returncode,)
+            else:
+                signum = result.returncode - 128
+                try:
+                    signame = signal.Signals(signum).name
+                except:
```
Comment (ignatenkobrain) [inline mesonbuild/mtest.py]: __[Flake8]__

[E722] do not use bare except'

**Comment by krader1961** [comment]
Hopefully someone with more experience working on this project will add appropriate unit tests. I manually tested this on macOS 10.12 and Ubuntu 16.04.

**Comment by nirbheek** [comment]
See: `run_unittests.py`. You will want to make a test under `test cases/unit` that raises, for instance, `SIGABORT`, and then make a unit test under the `AllPlatformTests` class that reads the test log and ensures that it has the correct contents.

There are other unit tests that do similar things, so you should be able to get inspiration from reading them :)

File: mesonbuild/mtest.py [hunk 2]
```diff
@@ -405,11 +405,32 @@ def print_stats(self, numlen, tests, name, result, i):
...
+        status = ''
+        if result.res is TestResult.FAIL:
+            # Note: We can't use `os.WIFSIGNALED(result.returncode)` and the
+            # related functions here because the status returned by subprocess
+            # is munged. It is supposed to return a negative value if the
+            # process was killed by a signal but in reality it returns a
+            # positive value with 128 added to the signal number. Which is the
+            # opposite of the usual UNIX exit status semantics. That is, a
+            # process killed by a signal typically has a returncode < 128 while
+            # a termination due to an `exit()` call has a returncode > 128. So
```
Comment (jpakkane) [inline mesonbuild/mtest.py]: Aren't these < and > characters the wrong way around in this particular sentence? They don't seem to match the text above or the code below. (that is, a process killed by a signal has a return code _larger_ than 128)

Comment (jon-turney) [reply mesonbuild/mtest.py]: I'm skeptical of the correctness of that whole comment .

Exit codes are unsigned integers (maybe even only 8 bit unsigned integers).  signals are not errnos. 

os.WTERMSIG/os.WIFSIGNALED should be the right tool to use here...

Comment (krader1961) [reply mesonbuild/mtest.py]: Gah! I just took my own advice. The python3 `subprocess` class behaves as documented with respect to the `returncode` attribute when I use it from Ipython or a hand crafted python program. Which implies Meson is munging the subprocess `returncode`. I'm going to withdraw this PR while I ponder the Meson code to understand why my change produces the expected result even though it is at odds with the documented behavior of the underlying `subprocess` module.

Comment (krader1961) [reply mesonbuild/mtest.py]: @jon-turney I just read your comment a second time and I don't understand what you mean by "signals are not errnos". Where in this PR did I suggest that errno values are relevant?

Comment (jon-turney) [reply mesonbuild/mtest.py]: Apologies, I hadn't noticed that sentence in the documentation for `subprocess`, and indeed [it does crack open the exit status for you](https://github.com/python/cpython/blob/3.6/Lib/subprocess.py#L1348-L1363), so you can just ignore what I wrote :smile:

Then I'd misunderstood what you wrote to say that wait() returns -signal, as some libc functions return -errno.

**Comment by jon-turney** [comment]
I'm not sure how you are testing this, with a simple test case, this seems to behave as documented, returning a negative value for a signal.

See https://github.com/jon-turney/meson/commit/85cc26e5e8e4d411b147f9aafbbd30e468f305c1

```console
jon@wollstonecraft:~/src/meson/test cases/failing test/1 trivial$ ~/src/meson/meson.py _build
The Meson build system
Version: 0.47.0.dev1
Source dir: /home/jon/src/meson/test cases/failing test/1 trivial
Build dir: /home/jon/src/meson/test cases/failing test/1 trivial/_build
Build type: native build
Project name: trivial
Project version: undefined
Native C compiler: cc (gcc 7.3.0 "cc (Ubuntu 7.3.0-16ubuntu3) 7.3.0")
Build machine cpu family: x86
Build machine cpu: i686
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
jon@wollstonecraft:~/src/meson/test cases/failing test/1 trivial$ ninja -C _build -v test
ninja: Entering directory `_build'
[1/5] cc -Imain@exe -I. -I.. -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g  -MD -MQ 'main@exe/main.c.o' -MF 'main@exe/main.c.o.d' -o 'main@exe/main.c.o' -c ../main.c
[2/5] cc -Isignal@exe -I. -I.. -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g  -MD -MQ 'signal@exe/signal.c.o' -MF 'signal@exe/signal.c.o.d' -o 'signal@exe/signal.c.o' -c ../signal.c
[3/5] cc  -o main 'main@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed
[4/5] cc  -o signal 'signal@exe/signal.c.o' -Wl,--no-undefined -Wl,--as-needed
[4/5] /usr/bin/python3 -u /home/jon/src/meson/meson.py test --no-rebuild --print-errorlogs
1/2 My Test                                 FAIL     0.02 s (exit status 1)
2/2 My Test Signal                          FAIL     0.02 s (killed by signal 2 SIGINT)
```

Is there a shell, or something else involved, which converts the terminating signal to exit status 128 + signal?

**Comment by nirbheek** [comment]
> I'm not sure how you are testing this, with a simple test case, this seems to behave as documented, returning a negative value for a signal.

Yep, I tested it as recommended above:

```py
>>> import subprocess
>>> p = subprocess.run(['/bin/sleep', '9999'])
```
```sh
$ killall sleep
```
```py
>>> p.returncode
-15
```

File: mesonbuild/mtest.py (hunk 2 above)
Comment (nirbheek) [reply mesonbuild/mtest.py]: > Which implies Meson is munging the subprocess returncode

There should be nothing in the Meson code that intentionally does that. If it happens, it's a bug.

**Comment by nirbheek** [comment]
This might be of interest: https://blogs.gnome.org/wjjt/2018/06/08/when-is-an-exit-code-not-an-exit-code/

**Comment by krader1961** [comment]
Yep, I was going to post essentially the same thing last night but instead spent some time swearing as I pondered the ksh code. See also https://stackoverflow.com/questions/1101957/are-there-any-standard-exit-status-codes-in-linux. I know the magic values of 126 and 127 are required by POSIX. I don't recall if POSIX requires or simply recommends that shells also set the exit status to the signal number plus 128. I tested bash, fish, ksh, and zsh. They all do the signo + 128 mapping of child processes that die due to a signal.

So what I'm going to do is rework my PR to distinguish between three cases:

1) Negative value, per the subprocess API, unambiguously means the command being tested died from a signal. For example, `kill by signal SIGSEGV`.

2) Positive value less than or equal to 128 is a normal exit status. For example, `exit status 3`.

3) Positive values greater than 128 may be a normal exit status or a signal number when 128 is subtracted. For example, `exit status 134 or SIGABRT`.

**Comment by krader1961** [comment]
I also took a look at how the ksh unit tests are run and confirmed they have a test runner script that sits between Meson and the actual unit test. That accounts for the conversion of failures like SIGSEGV to 128 + signo.

**Comment by jon-turney** [comment]
Continued in #3721

**Pull request closed without merging**